File: leetcode/easy/559_N叉树的最大深度/559_N叉树的最大深度.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

//递归法
func maxDepth(root *Node) int {
	if root == nil {
		return 0
	}

	depth := 0
	for _, child := range root.Children {
		childDepth := maxDepth(child)
		if childDepth > depth {
			depth = childDepth
		}
	}
	return depth + 1
}



//迭代法
func maxDepth(root *Node) int {
	if root == nil {
		return 0
	}

	depth := 0
	que := []*Node{root}
	for len(que) > 0 {
		size := len(que)
		depth++
		for i := 0; i < size; i++ {
			node := que[0]
			que = que[1:]
			for _, child := range node.Children {
				que = append(que, child)
			}
		}
	}
	return depth
}

File: leetcode/easy/559_N叉树的最大深度/559_N叉树的最大深度.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: Optional[int] = None, children: Optional[List['Node']] = None):
        self.val = val
        self.children = children
"""

#递归法
class Solution:
    def maxDepth(self, root: 'Node') -> int:
        if root is None:
            return 0

        depth = 0
        for child in root.children:
            depth = max(depth, self.maxDepth(child))
        return depth + 1



#迭代法
class Solution:
    def maxDepth(self, root: 'Node') -> int:
        if root is None:
            return 0

        depth = 0
        que = [root]
        while que:
            size = len(que)
            depth += 1
            for _ in range(size):
                node = que.pop(0)
                for child in node.children:
                    que.append(child)
        return depth